### PR TITLE
8251994: VM crashed running TestComplexAddrExpr.java test with -XX:UseAVX=X

### PR DESCRIPTION
**Pull request by vnkozlov**
To improve a chance to vectorize a loop, a special code in superword tries to hoist loads to the beginning of loop by replacing their memory input with corresponding (same memory slice) loop's memory Phi : 
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/superword.cpp#L474

Originally loads are ordered by corresponding stores on the same memory slice. But when they are hoisted they loose that ordering - nothing enforce the order. 
In TestComplexAddrExpr.test6 case the ordering is preserved (luckily?) after hoisting only when vector size is 32 bytes (avx2) but they become unordered with 16 (avx=0 or avx1) or 64 (avx512) bytes vectors. 

The mystery of why the test did not fail in our teting environment is also solved! We have old Skylake machines (even my local machine) for which AVX is switched to avx2: 
https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/vm_version_x86.cpp#L721

I have simple fix (use original loads ordering indexes) but looking on the code which causing the issue I see that it is bogus/incomplete - it does not help cases listed for JDK-8076284 changes: 

https://mail.openjdk.java.net/pipermail/hotspot-compiler-dev/2015-April/017645.html 

Using unrolling and cloning information to vectorize is interesting idea but as I see it is not complete. Even if pack_parallel() method is able created packs they are currently all removed by filter_packs() method. 
And additionally the above cases from JDK-8076284 are vectorized without hoisting loads and pack_parallel - I verified it.

The code added by  JDK-8076284  is useless now and I am excluding ti. It needs more work to be useful. I reluctant to remove the code because may be in a future we will have time to invest into it.

There were 2 way to exclude it: add new field in superword class:

```
   bool           _do_vector_loop;  // whether to do vectorization/simd style
+  bool           _do_vector_loop_experimental; // experimental optimization
   bool           _do_reserve_copy; // do reserve copy of the graph(loop) before final modification in output
```
or use `#if VECTOR_LOOP_SIMD` as in current changes. I prefer this one to avoid wasting compilation time. I used first one for testing by setting `_do_vector_loop_experimental(UseNewCode)`.

Testing: tier1-7 with avx2 and avx512.
Performance testing - no regrression.
I also compared jtreg tests output with -XX:+TraceNewVectors to verify that number of created vector nodes did not change.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x32 | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |  ✔️ (9/9 passed) | ⏳ (1/9 running) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8251994](https://bugs.openjdk.java.net/browse/JDK-8251994): VM crashed running TestComplexAddrExpr.java test with -XX:UseAVX=X

### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**)
 * [Claes Redestad](https://openjdk.java.net/census#redestad) (@cl4es - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/859/head:pull/859`
`$ git checkout pull/859`

### CODE REVIEW

File: src/hotspot/share/opto/superword.cpp [hunk 1]
```diff
@@ -1723,7 +1731,14 @@ void SuperWord::construct_my_pack_map() {
     Node_List* p = _packset.at(i);
     for (uint j = 0; j < p->size(); j++) {
       Node* s = p->at(j);
-      assert(my_pack(s) == NULL, "only in one pack");
+#ifdef ASSERT
+      if (my_pack(s) != NULL) {
+        s->dump(1);
+        tty->print_cr("packs[%d]:", i);
+        print_pack(p);
+        assert(false, "only in one pack");
```
Comment (shipilev) [inline src/hotspot/share/opto/superword.cpp:1739]: This is just `fatal`, right?

**Review by shipilev — CHANGES_REQUESTED** [review]
Drive-by comment: 
 - synopsis should be "crashed", not "crushed"?
 - I personally find `VECTOR_LOOP_SIMD` more fragile than the `_do_vector_loop_experimental` field. At least the macro should also say "experimental"?

File: src/hotspot/share/opto/superword.cpp (hunk 1 above)
Comment (vnkozlov) [reply src/hotspot/share/opto/superword.cpp:1739]: it is assert because this block of code is under `#ifdef ASSERT`

**Comment by vnkozlov** [comment]
> Drive-by comment:
> 
> * synopsis should be "crashed", not "crushed"?

Fixed.

> * I personally find `VECTOR_LOOP_SIMD` more fragile than the `_do_vector_loop_experimental` field. At least the macro should also say "experimental"?

'fragile' is in naming sense to use `DO_VECTOR_LOOP_EXPERIMENTAL`, for example?
Or you prefer to have the code be guarded by `if(_do_vector_loop_experimental)` runtime check instead of `#if DO_VECTOR_LOOP_EXPERIMENTAL` macro check?

**Comment by shipilev** [comment]
> 'fragile' is in naming sense to use `DO_VECTOR_LOOP_EXPERIMENTAL`, for example?

This makes most sense, I believe.

> Or you prefer to have the code be guarded by `if(_do_vector_loop_experimental)` runtime check instead of `#if DO_VECTOR_LOOP_EXPERIMENTAL` macro check?

Yeah, I am not a big fan of doing macros when runtime checks carry the same weight. I understand you want to micro-optimize these paths with macros, but I think the runtime checks work without much penalty here? I would defer to compiler reviewers to say which one is better.

**Comment by cl4es** [comment]
Wouldn't something like a `const bool _do_vector_loop_experimental = false;` eliminate the disabled code anyway?

**Comment by vnkozlov** [comment]
> Wouldn't something like a `const bool _do_vector_loop_experimental = false;` eliminate the disabled code anyway?

Good idea. I assume you mean `static const bool`. I will update changes.

**Review by shipilev — APPROVED** [review]
That is much cleaner, thanks. Looks good to me.

**Review by cl4es — APPROVED** [review]

**Comment by vnkozlov** [comment]
Thank you Aleksey and Claes for reviews.

File: src/hotspot/share/opto/superword.cpp [hunk 2]
```diff
@@ -91,6 +91,8 @@ SuperWord::SuperWord(PhaseIdealLoop* phase) :
 #endif
 }
 
+static const bool _do_vector_loop_experimental = false; // Experimental vectorization which uses data from loop unrolling.
```
Comment (iwanowww) [inline src/hotspot/share/opto/superword.cpp:94]: For such purposes, I find debug VM flags a better option: they are constant in product binaries, easy to experiment/change from command-line with debug binaries, and enumerated in a single place.

Any reason to prefer `static const` declarations?

Comment (vnkozlov) [reply src/hotspot/share/opto/superword.cpp:94]: Adding any flag is exception - we should avoid it if it is not needed immediately. We have a lot of flags already.
As I said currently this code is broken - I want to remove it from debug builds too.
And it is easy to modify this value later - make it a field with the same name and initialize with diagnostic UseNewCode flag value as I did for testing.

**Pull request closed without merging**